In [125]:
# Дан текст на естественном языке. Требуется найти границы предложений и сравнить скорость
# поиска предложения по номеру при помощи прямого поиска по тексту и при помощи индекса.


import re
sentences = []
string = 'Приведем два-три примера зверского, безбожного отношения к приговоренным, хотя бы для того, чтобы расстроить нервы супругам королевских прокуроров.'
list_of_strings = []
with open('Posledniy-den-prigovorennogo-k-smerti.txt') as file:
    text = file.readlines()
for l in text:
    sent = re.findall('\w+.*?[.?!]', l)
    for s in sent:
        sentences.append(s)

In [126]:
def linear_search(sentences, string):
    for t in sentences:
        if t == string:
            return True, sentences.index(t)

In [127]:
%%time
print(linear_search(sentences, string))

(True, 162)
CPU times: user 344 µs, sys: 179 µs, total: 523 µs
Wall time: 506 µs


In [128]:
string_length = len(string)
sent_lens = []
for l in sentences:
    sent_lens.append(len(l))
sent_lens.sort()

In [129]:
def index_search(string_length, sent_lens):
    mid = len(sent_lens) // 2
    low = 0
    high = len(sent_lens) - 1
    while sent_lens[mid] != string_length and low <= high:
        if string_length > sent_lens[mid]:
            low = mid + 1
        else:
            high = mid - 1
        mid = (low + high) // 2
    if low > high:
        return "No value"
    else:
        return "ID =", mid

In [130]:
%%time
print(index_search(string_length, sent_lens))

('ID =', 2233)
CPU times: user 219 µs, sys: 95 µs, total: 314 µs
Wall time: 265 µs


# Если элемент находится где-то в начале:

### Время выполнения линейного поиска по списку:
CPU times: user 219 µs, sys: 81 µs, total: 300 µs
Wall time: 249 µs
### Время выполнения поиска по индексу в отсортированном списке:
CPU times: user 332 µs, sys: 147 µs, total: 479 µs
Wall time: 374 µs

# Если элемент находится где-то в конце:

### Время выполнения линейного поиска по списку:
CPU times: user 615 µs, sys: 148 µs, total: 763 µs
Wall time: 670 µs
### Время выполнения поиска по индексу в отсортированном списке:
CPU times: user 303 µs, sys: 103 µs, total: 406 µs
Wall time: 358 µs

*в общем, если нам повезло и элемент находится в начале списка, то мы немного выигрываем на линейном поиске. Но если нам не повезло и элемент находится в конце списка, то мы сильно (в два раза) выигрываем на поиске по индексу.